src/app/DeferredAttributePersistenceProvider.cpp

```diff
@@ -0,0 +1,70 @@
+/*
+ *    Copyright (c) 2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <app/DeferredAttributePersistenceProvider.h>
+
+#include <platform/CHIPDeviceLayer.h>
+
+namespace chip {
+namespace app {
+
+CHIP_ERROR DeferredAttribute::PrepareWrite(AttributePersistenceProvider & persister, const EmberAfAttributeMetadata * metadata,
+                                           const ByteSpan & value)
+{
+    mPersister = &persister;
+    mMetadata  = metadata;
+
+    if (mValue.AllocatedSize() != value.size())
+    {
+        mValue.Alloc(value.size());
+        ReturnErrorCodeIf(!mValue, CHIP_ERROR_NO_MEMORY);
+    }
+
+    memcpy(mValue.Get(), value.data(), value.size());
+    return CHIP_NO_ERROR;
+}
+
+void DeferredAttribute::Flush()
+{
+    VerifyOrReturn(mValue);
+    mPersister->WriteValue(mPath, mMetadata, ByteSpan(mValue.Get(), mValue.AllocatedSize()));
```

Comment:
I'm a little worried about the metadata pointer here.  What happens if the metadata is for a dynamic endpoint that is removed, and has its metadata deallocated, between when the initial write happens and when the timer fires?

It looks like we mostly pass the metadata to WriteValue for symmetry with ReadValue, and in case it's useful to some persistence implementation.  But `DefaultAttributePersistenceProvider::WriteValue` never uses it.  I wonder whether it makes sense to just pass null here, not store the metadata, and document in `DefaultAttributePersistenceProvider::WriteValue` that if it ever starts using the metadata this part will need to change.

Reply:
Right, I haven't thought about dynamic endpoints. I can follow your suggestion. Alternatively, maybe we could remove metadata argument from `WriteValue` and that way if it ever gets needed, one will be force to fix all implementations?

Reply:
Removing it for now, since we have no obvious use cases, seems ok to me.